### PR TITLE
Add util function to fetch SDOs of a slave

### PR DESCRIPTION
**Pull request by flosse**

### CODE REVIEW

File: src/util.rs [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    slave_pos: SlavePos,
+) -> Result<HashMap<SdoIdx, SdoEntryInfo>> {
+    let slave = master.get_slave_info(slave_pos)?;
+    let sdo_positions = (0..slave.sdo_count).into_iter().map(SdoPos::from);
+    let mut res = HashMap::new();
+    for sdo_pos in sdo_positions {
+        let SdoInfo {
+            idx, max_sub_idx, ..
+        } = master.get_sdo(slave_pos, sdo_pos)?;
+        let sdo_entries = (0..=u8::from(max_sub_idx))
```
Comment (birkenfeld) [inline src/util.rs]: Isn't subindex 0 the max-subindex again?

Also, in my experience not all subindices up to max-subindex need to be valid. Is that handled here?

File: src/util.rs [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+use std::collections::HashMap;
+
+type Result<T> = std::result::Result<T, Error>;
+
+pub fn slave_sdos(
+    master: &mut Master,
+    slave_pos: SlavePos,
+) -> Result<HashMap<SdoIdx, SdoEntryInfo>> {
+    let slave = master.get_slave_info(slave_pos)?;
+    let sdo_positions = (0..slave.sdo_count).into_iter().map(SdoPos::from);
```
Comment (birkenfeld) [inline src/util.rs:11]: `into_iter()` should be unneeded since range is already an iterator. (Also below)

File: src/util.rs [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    let mut res = HashMap::new();
+    for sdo_pos in sdo_positions {
+        let SdoInfo {
+            idx, max_sub_idx, ..
+        } = master.get_sdo(slave_pos, sdo_pos)?;
+        let sdo_entries = (0..=u8::from(max_sub_idx))
+            .into_iter()
+            .map(SubIdx::from)
+            .map(|sub_idx| SdoIdx { idx, sub_idx })
+            .map(|sdo_idx| (sdo_idx, SdoEntryAddr::ByIdx(sdo_idx)))
```
Comment (birkenfeld) [inline src/util.rs]: Is that separate map really needed? why not construct `SdoEntryAddr` in the call for `get_sdo_entry`?

File: src/util.rs [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+            .map(|(sdo_idx, addr)| {
+                master
+                    .get_sdo_entry(slave_pos, addr)
+                    .map(|entry| (sdo_idx, entry))
+            })
+            .collect::<Result<Vec<_>>>()?;
```
Comment (birkenfeld) [inline src/util.rs]: I guess there's no need to create a Vec just to merge it into the hashmap immediately after...

File: src/util.rs (hunk 1 above)
Comment (flosse) [reply src/util.rs]: > Isn't subindex 0 the max-subindex again?

ok, so now I have
```rust
(0..u8::from(max_sub_idx))
```

> Also, in my experience not all subindices up to max-subindex need to be valid. Is that handled here?

no, I can't handle it ... I can just early return on errors.
Do you have an other idea?

Comment (birkenfeld) [reply src/util.rs]: Sorry, I was imprecise. I meant that (I thought that) querying subindex 0 gives you the `max_sub_idx`.  So you'd have to do `1..=max`?

As for handling errors, maybe there is a special error for the "not defined" case which could be handled by not returning early?  Sorry, can't test with my system right now.

Comment (flosse) [reply src/util.rs]: > Sorry, I was imprecise. I meant that (I thought that) querying subindex 0 gives you the `max_sub_idx`. So you'd have to do `1..=max`?

ok, done. 

> As for handling errors, maybe there is a special error for the "not defined" case which could be handled by not returning early? Sorry, can't test with my system right now.

ok, I'll try to test it soon with some real hardware.

Comment (flosse) [reply src/util.rs]: Within the original CLI they start at `0`:

```cpp
 for (i = 0; i < slave.sdo_count; i++) {
        m.getSdo(&sdo, slave.position, i);
```

http://hg.code.sf.net/p/etherlabmaster/code/file/tip/tool/CommandSdos.cpp#l136

Comment (birkenfeld) [reply src/util.rs]: Ok, I guess the easiest is to test with real hardware then.

**Comment by birkenfeld** [comment]
Closing, probably out of date.

**Pull request closed without merging**